examples/protyping/auto-segmentation/dicom_process.py

```diff
@@ -0,0 +1,383 @@
+# -*- coding: utf-8 -*-
+"""
+
+Document:
+This script aims to take dicom files and prepare them for:
+a) Training / Test data for the model (includes label processing)
+b) Prediction input for the model (model returns the labels)
+
+Example:
+
+Attributes:
+
+Todo:
+"""
+
+from __future__ import print_function
+
+from glob import glob
+
+import numpy as np
+
+import skimage.draw
+import skimage.transform
+
+import pydicom
+
+
+def list_files(data_path, ext=None):
+    """Returns a sorted list of all files names in data_path with (optional) ext.
+
+    Parameters
+    ----------
+    data_path : str
+        Path to list files
+    ext : str
+        Option to list only files with extension .ext
+
+    Returns
+    -------
+    file_names : list
+        Sorted list of file names
+    """
+
+    if ext is None:
+        file_names = glob(data_path + "/*")
+    else:
+        file_names = glob(data_path + "/*" + ext)
+    file_names.sort()
+    return file_names
+
+
+def read_dicom_files(file_names):
+    """
+    Returns a list of DICOM files read in from each file_name
+    in list file_names.
+    """
+    return [pydicom.dcmread(file_name, force=True) for file_name in file_names]
```

Comment:
Interestingly, if you wanted to, you could actually take advantage of `stop_before_pixels`:

https://pydicom.github.io/pydicom/stable/reference/generated/pydicom.filereader.dcmread.html?highlight=header#pydicom.filereader.dcmread

What you can do, is read only the header for all the DICOM files, work out which ones are structures, and then initially, just read those. Then, if you wanted to, you could then make it so that you only read the pixel data from a given CT slice if it actually had pixel data relevant to a slice within a structure set.

* I could imagine a neat workflow would be, load all DICOM header data only (not pixel data)
* Pick the structure you care about, extract the coords
* Select a slice within the structures
* Define which slices are then needed to be loaded into memory from the hard disk by going +/- however many slices are required for your UNet layers (each layer drops two slices, one superior, and one inferior (https://en.wikipedia.org/wiki/Anatomical_terms_of_location#Superior_and_inferior)).
* Then look up the list of CT headers that you have recorded, find which files correspond to those slices you need
* Then only load those CT DICOM files and no more.

Reply:
After digging further I believe something resembling this process would be a good idea.

In regards to our conversation about time spent processing data, I have estimated the following:

**Saving all slices as .npy**
ie. pairs of image_slice.npy and mask_volume_for_image_slice.npy
Time per patient (averaged): 21 s
Time for 1000 patients (estimated): 5 h 51 m
Disk usage for 1000 patients (estimated): 567 G

The statistics for TIFF are very similar. 

**Dicom format**
Disk usage for 1000 patients (estimated):  80 G

It seems like building our own generator function to handle dicom files will pay off in the future.  Especially if we can cache the structure files for each batch.

Reply:
Sounds like a good plan :). Also I think once you've got a handle on it the logic won't be too difficult. The hardest bit I imagine will be wrangling with TensorFlow Dataset objects within those pipeline functions.